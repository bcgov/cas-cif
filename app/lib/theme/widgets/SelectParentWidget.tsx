import Dropdown from "@button-inc/bcgov-theme/Dropdown";
import { useEffect, useMemo, useState } from "react";
import { WidgetProps } from "@rjsf/core";

export interface EntitySchema {
  list: [{ rowId: any; year?: number }];
  displayField: string;
  placeholder: string;
  label: string;
}

interface SelectParentComponentProps extends WidgetProps {
  parent: EntitySchema;
  child: EntitySchema;
  foreignKey: string;
  displayIfChildrenEmpty: any;
}

const SelectParentWidget: React.FunctionComponent<
  SelectParentComponentProps
> = (props) => {
  const {
    id,
    onChange,
    required,
    uiSchema,
    value,
    parent,
    child,
    foreignKey,
    displayIfChildrenEmpty,
  } = props;

  const parentValue =
    child.list.find((opt) => opt.rowId == parseInt(value))?.[foreignKey] || "";
  const [selectedParentId, setSelectedParentId] = useState(parentValue);
  useEffect(() => {
    setSelectedParentId(parentValue);
  }, [parentValue]);

  const onParentChange = (val) => {
    setSelectedParentId(parseInt(val));
    onChange(parseInt(val));
    if (!parseInt(val)) onChange(undefined);
  };

  const childOptions = useMemo(() => {
    return child.list.filter((opt) => {
      return opt[foreignKey] === selectedParentId;
    });
  }, [child, foreignKey, selectedParentId]);

  return (
    <div>
      <label htmlFor={`select-parent-dropdown-${id}`}>{parent.label}</label>
      <Dropdown
        id={`select-parent-dropdown-${id}`}
        onChange={(e) => onParentChange(e.target.value || undefined)}
        size={(uiSchema && uiSchema["bcgov:size"]) || "large"}
        required={required}
        value={selectedParentId}
      >
        <option key={`option-placeholder-${id}`} value={undefined}>
          {parent.placeholder}
        </option>
        {parent.list.map((opt) => {
          return (
            <option key={opt.rowId} value={opt.rowId}>
              {opt[parent.displayField]}
            </option>
          );
        })}
      </Dropdown>

      <label htmlFor={`select-child-dropdown-${id}`}>{child.label}</label>
      <Dropdown
        id={`select-child-dropdown-${id}`}
        onChange={(e) => onChange(parseInt(e.target.value) || undefined)}
        size={(uiSchema && uiSchema["bcgov:size"]) || "large"}
        required={required}
        value={value}
      >
        {child.placeholder && (
          <option key={`option-placeholder-${id}`} value={undefined}>
            {child.placeholder}
          </option>
        )}
        {displayIfChildrenEmpty && childOptions.length === 0 ? (
          <option>{displayIfChildrenEmpty}</option>
        ) : (
          childOptions.map((opt) => {
            return (
              <option key={opt.rowId} value={opt.rowId}>
                {opt[child.displayField]}
              </option>
            );
          })
        )}
      </Dropdown>
      <style jsx>
        {`
          div :global(input) {
            width: 100%;
          }
          div :global(.pg-select-wrapper) {
            padding: 9px 2px 9px 2px;
            margin-bottom: 5px;
            margin-top: 2px;
          }
        `}
      </style>
    </div>
  );
};

export default SelectParentWidget;
