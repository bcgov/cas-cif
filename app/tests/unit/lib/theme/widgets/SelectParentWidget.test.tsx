import React from "react";
import SelectParentWidget from "lib/theme/widgets/SelectParentWidget";
import { render, fireEvent, screen, act } from "@testing-library/react";

const onChangeMock = jest.fn();

const props: any = {
  id: "test-id",
  onChange: onChangeMock,
  value: undefined,
  isInternal: true,
  required: false,
  parent: {
    list: [
      {
        rowId: 1,
        name: "EP",
        fundingStreamId: 1,
      },
      {
        rowId: 2,
        name: "IA",
        fundingStreamId: 2,
      },
    ],
    label: "Parent Label",
    placeholder: "parent placeholder",
    displayField: "name",
  },
  child: {
    list: [
      {
        rowId: 1,
        year: 2019,
        fundingStreamId: 1,
      },
      {
        rowId: 2,
        year: 2020,
        fundingStreamId: 1,
      },
      {
        rowId: 3,
        year: 2021,
        fundingStreamId: 2,
      },
      {
        rowId: 4,
        year: 2022,
        fundingStreamId: 2,
      },
    ],
    label: "Child Label",
    placeholder: "child placeholder",
    displayField: "year",
  },
  foreignKey: "fundingStreamId",
};

describe("The SelectParentWidget Widget", () => {
  it("selects the first option from parent", async () => {
    render(<SelectParentWidget {...props} />);
    const parentDropdown = screen.getByLabelText("Parent Label");
    expect(parentDropdown.value).toEqual("parent placeholder");
    act(() => fireEvent.change(parentDropdown, { target: { value: "1" } }));
    expect(parentDropdown.value).toEqual("1");
    expect(onChangeMock).toHaveBeenCalledWith(undefined);
  });

  it("selects the first option from child", async () => {
    render(<SelectParentWidget {...props} />);
    const childDropdown = screen.getByLabelText("Child Label");
    const parentDropdown = screen.getByLabelText("Parent Label");
    fireEvent.change(parentDropdown, { target: { value: "2" } });
    expect(childDropdown.value).toEqual("child placeholder");
    fireEvent.change(childDropdown, { target: { value: "3" } });
    expect(childDropdown.value).toEqual("3");
  });

  it("selects the first option from parent by an external user", async () => {
    render(<SelectParentWidget {...props} isInternal={false} />);
    const parentDropdownExternal = screen.getByLabelText("Parent Label");
    act(() =>
      fireEvent.change(parentDropdownExternal, { target: { value: "1" } })
    );
    expect(parentDropdownExternal.value).toEqual("1");
    // calls the rowId of the latest option by year in the child list that matches the parent value
    expect(onChangeMock).toHaveBeenCalledWith(2);
  });

  it("cannot select the first option from child with non-selected parent", async () => {
    render(<SelectParentWidget {...props} />);
    const childDropdown = screen.getByLabelText("Child Label");
    const parentDropdown = screen.getByLabelText("Parent Label");
    fireEvent.change(parentDropdown, { target: { value: "2" } });
    expect(childDropdown.value).toEqual("child placeholder");
    fireEvent.change(childDropdown, { target: { value: "1" } });
    expect(childDropdown.value).toEqual("");
  });

  it("Matches the snapshot with a default value set", () => {
    const componentUnderTest = render(<SelectParentWidget {...props} />);
    expect(componentUnderTest.container).toMatchSnapshot();
  });

  it("does not show a child placeholder if none is provided", async () => {
    const propsWithoutChildPlaceholder: any = {
      id: "test-id",
      onChange: onChangeMock,
      value: undefined,
      required: false,
      parent: {
        list: [
          {
            rowId: 1,
            name: "EP",
            fundingStreamId: 1,
          },
          {
            rowId: 2,
            name: "IA",
            fundingStreamId: 2,
          },
        ],
        label: "Parent Label",
        placeholder: "parent placeholder",
        displayField: "name",
      },
      child: {
        list: [
          {
            rowId: 1,
            year: 2019,
            fundingStreamId: 1,
          },
          {
            rowId: 2,
            year: 2020,
            fundingStreamId: 1,
          },
          {
            rowId: 3,
            year: 2021,
            fundingStreamId: 2,
          },
          {
            rowId: 4,
            year: 2022,
            fundingStreamId: 2,
          },
        ],
        label: "Child Label",
        displayField: "year",
      },
      foreignKey: "fundingStreamId",
    };
    render(<SelectParentWidget {...propsWithoutChildPlaceholder} />);
    const childDropdown = screen.getByLabelText("Child Label");
    expect(childDropdown.value).toBe("");
  });
});
