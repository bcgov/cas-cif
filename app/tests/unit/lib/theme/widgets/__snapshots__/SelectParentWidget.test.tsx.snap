// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The SelectParentWidget Widget Matches the snapshot with a default value set 1`] = `
<div>
  <div
    class="jsx-3921837838"
  >
    <label
      class="jsx-3921837838"
      for="select-parent-dropdown-test-id"
    >
      Parent Label
    </label>
    <div
      class="sc-eCYdqJ bJnssn pg-select"
    >
      <div
        class="sc-gKXOVf kjA-DCE pg-select-wrapper"
      >
        <select
          aria-label="select-parent-dropdown-test-id-select"
          class="sc-iBkjds fhVste pg-select-input"
          id="select-parent-dropdown-test-id"
          name="select-parent-dropdown-test-id-select"
        >
          <option
            class="jsx-3921837838"
          >
            parent placeholder
          </option>
          <option
            class="jsx-3921837838"
            value="1"
          >
            EP
          </option>
          <option
            class="jsx-3921837838"
            value="2"
          >
            IA
          </option>
        </select>
      </div>
    </div>
    <label
      class="jsx-3921837838"
      for="select-child-dropdown-test-id"
    >
      Child Label
    </label>
    <div
      class="sc-eCYdqJ bJnssn pg-select"
    >
      <div
        class="sc-gKXOVf kjA-DCE pg-select-wrapper"
      >
        <select
          aria-label="select-child-dropdown-test-id-select"
          class="sc-iBkjds fhVste pg-select-input"
          id="select-child-dropdown-test-id"
          name="select-child-dropdown-test-id-select"
        >
          <option
            class="jsx-3921837838"
          >
            child placeholder
          </option>
        </select>
      </div>
    </div>
  </div>
</div>
`;
